src/urllib3/multipart/decoder.py

```diff
@@ -0,0 +1,149 @@
+        return self.content.decode(self.encoding)
+
+
+MD = typing.TypeVar("MD", bound="MultipartDecoder")
+
+
+class MultipartDecoder:
+    """This parses the full multipart/form-data payload.
+
+    The ``MultipartDecoder`` object parses the multipart payload of
+    a bytestring into a tuple of ``Response``-like ``BodyPart`` objects.
+
+    The basic usage is::
+
+        import requests
+        from requests_toolbelt import MultipartDecoder
+
+        response = request.get(url)
+        decoder = MultipartDecoder.from_response(response)
+        for part in decoder.parts:
+            print(part.headers['content-type'])
+
+    If the multipart content is not from a response, basic usage is::
+
+        from requests_toolbelt import MultipartDecoder
+
+        decoder = MultipartDecoder(content, content_type)
+        for part in decoder.parts:
+            print(part.headers['content-type'])
+
+    For both these usages, there is an optional ``encoding`` parameter. This is
+    a string, which is the name of the unicode codec to use (default is
+    ``'utf-8'``).
+    """
+
+    def __init__(self, content: bytes, content_type: str, encoding: str = "utf-8"):
+        #: Original Content-Type header
+        self.content_type = content_type
+        #: Response body encoding
+        self.encoding = encoding
+        #: Parsed parts of the multipart response body
+        self.parts: tuple[BodyPart, ...] = tuple()
+        self._find_boundary()
+        self._parse_body(content)
+
+    def _find_boundary(self) -> None:
+        ct_info = tuple(x.strip() for x in self.content_type.split(";"))
+        mimetype = ct_info[0]
+        if mimetype.split("/")[0].lower() != "multipart":
+            raise NonMultipartContentTypeError(
+                f"Unexpected mimetype in content-type: '{mimetype}'"
+            )
+        for item in ct_info[1:]:
+            attr, _, value = item.partition("=")
+            if attr.lower() == "boundary":
+                self.boundary = encode_with(value.strip('"'), self.encoding)
+
+    @staticmethod
+    def _fix_first_part(part: bytes, boundary_marker: bytes) -> bytes:
+        bm_len = len(boundary_marker)
+        if boundary_marker == part[:bm_len]:
+            return part[bm_len:]
+        else:
+            return part
+
+    def _parse_body(self, content: bytes) -> None:
+        boundary = b"--" + self.boundary
+
+        def body_part(part: bytes) -> BodyPart:
+            fixed = MultipartDecoder._fix_first_part(part, boundary)
+            return BodyPart(fixed, self.encoding)
+
+        def test_part(part: bytes) -> bool:
+            return (
+                part != b""
+                and part != b"\r\n"
+                and part[:4] != b"--\r\n"
+                and part != b"--"
+            )
+
+        parts = content.split(b"\r\n" + boundary)
+        self.parts = tuple(body_part(x) for x in parts if test_part(x))
+
+    @classmethod
+    def from_response(
+        cls: type[MD],
```

Comment:
Minor: elsewhere in the project, the `cls` parameter is left untyped and mypy infers the correct type.

e.g. https://github.com/urllib3/urllib3/blob/733f638a2faa02b4ff8a9f3b5668949d39396b8b/src/urllib3/util/retry.py#L266-L268